lib/ets/extra.ex

```diff
@@ -0,0 +1,86 @@
+defmodule ETS.Extra.Guards do
+  @moduledoc """
+  Guard macros for ETS tables.
+
+  """
+
+  @doc """
+  Returns true if the given term *could* refer to an ETS table.
+
+  Allowed in match guards.
+
+  """
+  defmacro is_table(term) do
+    quote do: is_atom(unquote(term)) or is_reference(unquote(term))
+  end
+end
+
+
+
+defmodule ETS.Extra do
+  @moduledoc """
+  Utility functions for dealing with ETS tables.
+
+  """
+
+  import ETS.Extra.Guards
+
+  @type table :: atom | reference
+
+
+  @doc """
+  Stream all keys from the given ETS table.
+
+  ## Examples
+
+      iex> table = :ets.new(:my_table, [])
+      iex> :ets.insert(table, {:foo, 1})
+      iex> :ets.insert(table, {:bar, 2})
+      iex> ETS.Extra.stream_keys(table) |> Enum.to_list()
+      [:bar, :foo]
+
+  """
+  @spec stream_keys(table) :: Enum.t
+  def stream_keys(table) when is_table(table) do
+    eot = :"$end_of_table"
+
+    Stream.resource(
+      fn -> [] end,
+      fn
+        [] ->
+          case :ets.first(table) do
+            ^eot -> {:halt, []}
+            first_key -> {[first_key], first_key}
+          end
+
+        acc ->
+          case :ets.next(table, acc) do
+            ^eot -> {:halt, acc}
+            next_key -> {[next_key], next_key}
+          end
+      end,
+
+      fn _acc -> :ok end
+    )
+  end
+
+
+  @doc """
+  Stream objects from the given ETS table as `{key, value}` pairs.
+
+  ## Examples
+
+      iex> table = :ets.new(:my_table, [])
+      iex> :ets.insert(table, {:foo, 1})
+      iex> :ets.insert(table, {:bar, 2})
+      iex> ETS.Extra.stream(table) |> Enum.to_list()
+      [bar: 2, foo: 1]
+
+  """
```

Comment:
It's worth noting that this implementation presumes a `:set` or `:ordered_set` ETS table, as opposed to a bag implementation, for which it will only return the first value.

At minimum we should note this, but realistically we should probably `Stream.flat_map` the inner elements to their originating keys.

ie. `[{:duplicate_key, "Item"}, {:duplicate_key, "Item 2"}]` would be emitted from the enumerable in a bag mode ETS table. Alternatively we might want to keep them as lists and let the caller deal with them (ie. I could see where, if I am streaming a bag, it is desirable to have the form of `{:duplicate_key, ["Item 1", "Item 2"]}`.

For more info on bags, see [the erlang docs](http://erlang.org/doc/man/ets.html)